dask/blockwise.py

```diff
@@ -1314,6 +1361,18 @@ def _optimize_blockwise(full_graph, keys=()):
                     new_deps |= keys_in_tasks(full_graph.dependencies, [k])
                 elif k not in io_names:
                     new_deps.add(k)
+                elif layers[layer].io_deps[k].valid:
+                    # Need to add valid key dependencies in io_deps[k].
+                    # Use required `numblocks` attribute to generate all possible input keys
+                    valid_keys = product(
+                        *(
+                            list(range(nbi))
+                            for nbi in layers[layer].io_deps[k].numblocks
+                        )
+                    )
+                    new_deps |= {
+                        layers[layer].io_deps[k][valid_key] for valid_key in valid_keys
+                    }
```

Comment:
Am I wrong to be thinking that a simpler way to get the set of `valid` deps would be to read them off of `full_graph.dependencies`?

Reply:
Grr - I'm still trying to work this one out. We certainly can pull out the valid dependencies from the intial version of `full_graph`, but this can become tricky by the time we get to the final version of `full_graph` (since this function is called *somewhat* recursively).

Reply:
Okay - I *think* (hope) [d7ca3f1](https://github.com/dask/dask/pull/8852/commits/d7ca3f18c328affbc96d248192315f8913072f9b) does what you had in mind.  (Although, I may need to clean up the exact code I used)

Reply:
Thanks @rjzamora, that is what I had in mind.

I'm not 100% sure it's better, but to me it *feels* more idiomatic to not explicitly iterate over the keys for a non-materialized graph at this stage. What do you think?

Reply:
I do get the sense that it is better. It also seems like it is the "correct" thing to do even without the new `produces_keys` feature.  For example, I am now realizing that we may be able to remove the block below it:

```python
                elif k not in io_names:
                    new_deps.add(k)
```
because these dependencies *should* already be in `new_deps` as well.

Reply:
Yup. I was able to just replace the original `new_deps` logic by pulling the dependencies straight from `full_graph.dependencies`. This seems like a nice improvement over the original `_optimize_blockwise` code anyway (hope CI passes).

Reply:
Ah, bummer. Thanks for checking